src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/Testing/RunTestsHandler.cs

```diff
@@ -0,0 +1,131 @@
+
+using System.Composition;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Testing;
+using Microsoft.TestPlatform.VsTestConsole.TranslationLayer;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Testing;
+
+[ExportCSharpVisualBasicStatelessLspService(typeof(RunTestsHandler)), Shared]
+[Method(RunTestsMethodName)]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal class RunTestsHandler(DotnetCliHelper dotnetCliHelper, TestDiscoverer testDiscoverer, TestRunner testRunner, ServerConfiguration serverConfiguration) : ILspServiceDocumentRequestHandler<RunTestsParams, RunTestsPartialResult[]>
+{
+    private const string RunTestsMethodName = "textDocument/runTests";
+
+    public bool MutatesSolutionState => false;
+
+    public bool RequiresLSPSolution => true;
+
+    public LSP.TextDocumentIdentifier GetTextDocumentIdentifier(RunTestsParams request)
+    {
+        return request.TextDocument;
+    }
+
+    public async Task<RunTestsPartialResult[]> HandleRequestAsync(RunTestsParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        Contract.ThrowIfNull(context.Document);
+        using var progress = BufferedProgress.Create(request.PartialResultToken);
+
+        // First, build to make sure we have a relatively up to date project.
+        await BuildAsync(context.Document, progress, cancellationToken);
+
+        var projectOutputPath = context.Document.Project.OutputFilePath;
+        Contract.ThrowIfFalse(File.Exists(projectOutputPath), $"Output path {projectOutputPath} is missing");
+
+        // Find the appropriate vstest.console.dll from the SDK.
+        var vsTestConsolePath = await dotnetCliHelper.GetVsTestConsolePathAsync(cancellationToken);
+
+        // Instantiate the test platform wrapper.
+        var vsTestConsoleWrapper = new VsTestConsoleWrapper(vsTestConsolePath, new ConsoleParameters
+        {
+            LogFilePath = Path.Combine(serverConfiguration.ExtensionLogDirectory, "testLogs", "vsTestLogs.txt"),
+            TraceLevel = GetTraceLevel(serverConfiguration),
+            EnvironmentVariables = new()
+            {
+                // Reset dotnet root to the user's original configuration so that vs test console can find the right runtimes.
+                { DotnetCliHelper.DotnetRootEnvVar, DotnetCliHelper.GetUserDotnetRoot() },
+            }
+        });
+
+        var testCases = await testDiscoverer.DiscoverTestsAsync(request.Range, context.Document, projectOutputPath, progress, vsTestConsoleWrapper, cancellationToken);
+        if (testCases.IsEmpty)
+        {
+            return progress.GetValues() ?? Array.Empty<RunTestsPartialResult>();
```

Comment:
is the `??` needed?

Reply:
Yeah - in the case of streaming the API is defined as being nullable.